core/corehttp/gateway_handler.go

```diff
@@ -415,11 +434,95 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
 		return
 	}
 
+	var querySelector ipld.Node
+	var selectorCID cid.Cid
+	if selectorParam := r.URL.Query().Get("selector"); selectorParam != "" {
+		selectorCID, err = cid.Decode(selectorParam)
+		if err != nil {
+			webError(w, "selector is not a valid CID", err, http.StatusInternalServerError)
+			return
+		}
+		selectorNode, err := i.api.Dag().Get(r.Context(), selectorCID)
+		if err != nil {
+			webError(w, "could not get selector", err, http.StatusInternalServerError)
+			return
+		}
+		querySelector = selectorNode.(ipld.Node)
+	}
+
+	if ipldPathParam := r.URL.Query().Get("ipld-path"); ipldPathParam != "" || ns == "ipld" {
+		var pcs []string
+		if ipldPathParam != "" {
+			pcs = strings.Split(ipldPathParam, "|")
+		} else {
+			pcs = strings.Split(contentPath.String(), "/")[3:]
+		}
+		buildSel := builder.NewSelectorSpecBuilder(basicnode.Prototype__Any{})
+		var fnBuildSel func(comps []string) (builder.SelectorSpec, error)
+		fnBuildSel = func(comps []string) (builder.SelectorSpec, error) {
```

Comment:
```suggestion
		fnBuildSel := func(comps []string) (builder.SelectorSpec, error) {
```

Reply:
Pretty sure the code won't compile if you do that since the function is recursive

Reply:
let me check

Reply:
Ok you are right, it's cursed to have a recursive virtual call to a function pointer if you aren't doing DP.

I would move this out as a private function on the global scope.